python/revng/daemon/__init__.py

```diff
@@ -93,19 +94,35 @@ def shutdown():
     capi_shutdown()
 
 
-app = Starlette(
-    debug=DEBUG,
-    middleware=[
+def get_middlewares() -> List[Middleware]:
+    extra_middlewares: List[Middleware] = []
+    if "STARLETTE_MIDDLEWARES" in os.environ:
+        for element in os.environ["STARLETTE_MIDDLEWARES"].split(","):
+            module_path, attribute_name = element.split(":", 1)
+            module = import_module(module_path)
+            if not hasattr(module, attribute_name):
```

Comment:
1. Can we go by convention? That is, force middlewares to implement a function/object whatever with a specific name.
2. Have you considered making a canonical path where to put middlewares so we don't need an additional `STARLETTE_MIDDLEWARES`.

Conventions are good!

Reply:
1. We can do that, but then it's a bit annoying if you have a package providing more than one middleware
2. That's possible but __really__ annoying, e.g. right now the auth middleware it's packaged and can declare dependencies and be cleanly packaged

Reply:
OK, let's import everything in a path automatically and then let's import but fully qualified name as django does.

Reply:
After having implemented `STARLETTE_MIDDLEWARE_{EARLY,LATE}` I'd really push just to use those, the special folder is just a special case and it can lead to unexpected surprises  later on

Reply:
For example?

In the end they'll end up one on the side of the other, why leave a degree of freedom we don't need?
We can have `share/revng/daemon/middlewares/` and be done with it.

Reply:
Because it's an incredibly poor way to handle them, it's really better to have each middleware/middleware group packaged as a proper python package that specifies its dependencies and it's installed the proper way via pip.
Also folders are just annoying, it usually leads to things like `<folder>/01_early_middleware.py` etc. and `<folder>/middleware.py.disabled`

Reply:
Having an environment variable which is a *list* of *pairs* of paths + names is not great either. Doing configuration in environment variables is already bad enough, let's not put JSONs in there please.

If you want, we can assume they are in `sys.path` and just have the environment variables specify the names.
No new magic paths, just some place already in `sys.path`.

Reply:
Right now there's no magic paths, only imports

e.g. the auth middleware is in `revng_middleware_auth.AuthMiddleware`, that's it
It's the equivalent of `from revng_middleware_auth import AuthMiddleware`

I know that comma-separated environment variables are not the greatest, but this is the most python way to handle things IMHO (and that would be extremely intuitive to anybody that has used a similar mechanism in the past)

Reply:
Then the middleware is already installed in someplace in `sys.path`?

Reply:
Yes